### PR TITLE
Add nanoseconds without dots

### PR DESCRIPTION
**Pull request by emschwartz**
Resolves https://github.com/chronotope/chrono/issues/250

### CODE REVIEW

**Comment by quodlibetor** [comment]
Looking great!

Regarding your questions:

* Yeah, I think we should duplicate the nanosecond parse tests, all of them. You can verify that you are only finding nanoseconds without dots by doing something like:

  ```rust
  check!(".400", [fix!(Nanosecond3NoDot)]; INVALID);  // probably not actually INVALID
  check!("400", [fix!(Nanosecond3NoDot)]; nanosecond: 400_000_000);
  ```

  Does that make sense?
* Only other place to update is [the strftime docs](https://github.com/emschwartz/chrono/blob/efafef42e30dd3a919d9be85dbf45c2dcc65947a/src/format/strftime.rs#L57-L60) and the CHANGELOG.md

**Comment by emschwartz** [comment]
Good catch. It was an issue with using the `try_consume` macro with a slice of the string. `try_consume` updates `s` in place, but if you pass a slice it forgets about the whole rest of the string

File: src/format/parse.rs [hunk 1]
```diff
@@ -532,6 +557,46 @@ fn test_parse() {
...
+
+    check!("",             [fix!(Nanosecond9NoDot)]; ); // no field set, but not an error
+    check!("0",            [fix!(Nanosecond9NoDot)]; nanosecond: 0);
+    check!("4",            [fix!(Nanosecond9NoDot)]; nanosecond: 400_000_000);
+    check!("42",           [fix!(Nanosecond9NoDot)]; nanosecond: 420_000_000);
+    check!("421",          [fix!(Nanosecond9NoDot)]; nanosecond: 421_000_000);
+    check!("42195",        [fix!(Nanosecond9NoDot)]; nanosecond: 421_950_000);
+    check!("421950803",    [fix!(Nanosecond9NoDot)]; nanosecond: 421_950_803);
+    check!("421950803547", [fix!(Nanosecond9NoDot)]; nanosecond: 421_950_803);
+    // check!("421950803547", [fix!(Nanosecond9NoDot), num!(Second)]; nanosecond: 421_950_803, second: 54); // don't skip digits that come after the 9
```
Comment (quodlibetor) [inline src/format/parse.rs]: did you want to uncomment this?

Comment (emschwartz) [reply src/format/parse.rs]: Yeah, I'm fixing the test. I'll add another commit to the PR tomorrow

**Comment by emschwartz** [comment]
Yeah, I'll make those changes tomorrow.

Would you prefer me to squash all the commits into one or leave them as they are?

**Comment by quodlibetor** [comment]
The commits are nicely separated as is, no need to squash. I would also appreciate the refactoring to use `InternalInternal` to be its own commit, I hope to revert it with the next breaking change.  

Speaking of it, though, if you are going to be editing your test commit *and* you use git from the command line I have [a *very* new project](https://github.com/quodlibetor/git-fixup) that I'm looking for beta testers/bug reporters/contributors on that is supposed to make editing old commits a bit nicer.

**Comment by emschwartz** [comment]
> Could you move your new variants into the InternalInternal struct, the equivalent of this commit for your changes? It didn't exist when you started work, but I think that, sadly, it's necessary for strict semver compat.

I'm not sure I understand that. Is the idea to make it so this feature can't be accessed by the user? What is the semver issue (it's a non-breaking change)?

**Comment by quodlibetor** [comment]
Adding new variants to a public enum [is a major breaking change](https://github.com/rust-lang/rfcs/blob/master/text/1105-api-evolution.md#enums). The idea isn't to prevent usage of the feature, it's to require that the only usage of the feature is through the `parse` method, specifically so that we don't break semver.

The hack used by the ecosystem to mark enums as open is to have a `#[doc(hidden)] __DoNotMatchAgainstMe` variant and to include in the docs for an enum that additions to it are expected. However, chrono was created before that convention existed, and so it created the separation of internal and external enums. The internal enum is not public, and so it can be expanded without trouble. The external enum, though, is public and so technically someone could be doing an exhaustive match against the whole thing.

For most libraries I think that just having bad docs would mean that you can just fix them and improve life, but chrono is one of the most depended-on crates in the ecosystem and so if any crate is going to have someone doing crazy things against massive enums then chrono might be it.

**Comment by quodlibetor** [comment]
specifically, the only things that you should need to change should be to move your variants to the `InternalInternal` enum and to change your uses of `fix!` to `internal_fix!`, all tests (in particular tests that verify `parse`) should continue to work without any other changes.

**Comment by emschwartz** [comment]
Gotcha. Thanks for the explanation. Should it be `InternalInternal` or `InternalFixed`?

**Comment by emschwartz** [comment]
Made the change. Ready for re-review

**Comment by quodlibetor** [comment]
LGTM, I'm going to do one more review pass this evening, I don't recall any other possible concerns so hopefully I'll get this published tonight.

**Comment by emschwartz** [comment]
Sounds good. Thanks for the help!

**Pull request merged**